app/src/main/java/com/kodeco/android/countryinfo/repository/CountryRepositoryImpl.kt

```diff
@@ -1,54 +1,60 @@
 package com.kodeco.android.countryinfo.repository
 
 import com.kodeco.android.countryinfo.model.Country
-import com.kodeco.android.countryinfo.networking.RemoteApiService
-import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.asStateFlow
+import com.kodeco.android.countryinfo.repository.local.CountryLocalDataSource
+import com.kodeco.android.countryinfo.repository.remote.CountryRemoteDataSource
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
 
 class CountryRepositoryImpl(
-    private val apiService: RemoteApiService,
+    private val remoteDataSource: CountryRemoteDataSource,
+    private val localDataSource: CountryLocalDataSource,
 ) : CountryRepository {
 
-    private val _countries = MutableStateFlow(emptyList<Country>())
-    override val countries = _countries.asStateFlow()
+    private var _countries = mutableListOf<Country>()
 
-    private var favorites = setOf<String>()
+    override fun fetchCountries(): Flow<List<Country>> = flow {
+        val countriesResponse = remoteDataSource.fetchCountries()
+        _countries.clear()
 
-    override suspend fun fetchCountries() {
-        val countriesResponse = apiService.getAllCountries()
-        _countries.value = emptyList()
+        if (countriesResponse.isSuccessful) {
+            countriesResponse.body()?.let {
+                deleteAllCountries()
+                saveCountries(it)
+                _countries.addAll(it)
 
-        _countries.value = try {
-            if (countriesResponse.isSuccessful) {
-                countriesResponse.body()!!
-                    .toMutableList()
-                    .map { country ->
-                        country.copy(isFavorite = favorites.contains(country.name.common))
-                    }
-            } else {
-                throw Throwable("Request failed: ${countriesResponse.message()}")
+                emit(it)
             }
-        } catch (e: Exception) {
-            throw Throwable("Request failed: ${e.message}")
+        } else {
+            throw Throwable("Request failed: ${countriesResponse.message()}")
         }
     }
 
-    override fun getCountry(name: String): Country? =
-        _countries.value.firstOrNull { country -> country.name.common == name }
+    override suspend fun saveCountries(countries: List<Country>) = localDataSource
+        .addCountry(*countries.toTypedArray())
 
-    override fun favorite(country: Country) {
-        favorites = if (favorites.contains(country.name.common)) {
-            favorites - country.name.common
-        } else {
-            favorites + country.name.common
+    override suspend fun deleteAllCountries() = localDataSource
+        .deleteAllCountries()
+
+    override fun getCountries(): Flow<List<Country>> = flow {
+        localDataSource.getAllCountriesFlow().collect {
+            _countries.clear()
+            _countries.addAll(it)
+            emit(it)
         }
+    }
```

Comment:
This can likely become private, based on my code suggestion above.  The idea with the offline mode feature is to be able to fetch first from local storage, then attempt to fetch from the remote data source.  The way the setup is now, when the local storage feature is enabled, the app _only_ fetches from local storage.  My code snippet above shows how you can have the repository reconcile local and network data.